Author: Kevin Chevreuil - Kaisen <kaisen@kaisenlinux.org>

--- partman-btrfs-54.orig/fstab.d/btrfs
+++ partman-btrfs-54/fstab.d/btrfs
@@ -12,21 +12,52 @@ for dev in $DEVICES/*; do
 		[ -f "$id/acting_filesystem" ] || continue
 		method=$(cat $id/method)
 		filesystem=$(cat $id/acting_filesystem)
+		mountpoint=$(cat $id/mountpoint)
 		case "$filesystem" in
 		    btrfs)
 			[ -f "$id/mountpoint" ] || continue
-			mountpoint=$(cat $id/mountpoint)
 			# due to #249322, #255135, #258117:
-			if [ "$mountpoint" = "/tmp" ]; then
+			if [ "$mountpoint" = /tmp ]; then
 				rm -f $id/options/noexec
 			fi
-			if [ "$mountpoint" = "/" ]; then
-				options="$(get_mountoptions $dev $id),subvol=@rootfs"
-			else
-				options=$(get_mountoptions $dev $id)
+			options=$(get_mountoptions $dev $id)
+			if [ $options = defaults ]; then
+				options=acl,noautodefrag,barrier,compress=zstd,datacow,datasum,discard=async,noflushoncommit,space_cache=v2,treelog,noatime,nodiratime
 			fi
-			# There is no btrfs fsck
-			echo "$path" "$mountpoint" btrfs $options 0 0
+			if [ $mountpoint = / ]; then
+				options="${options:+$options,}subvol=@"
+                        fi
+			if [ $mountpoint = /boot ]; then
+                                options="${options:+$options,}subvol=@boot"
+                        fi
+			if [ $mountpoint = /home ]; then
+				options="${options:+$options,}subvol=@home"
+			fi
+			if [ $mountpoint = /var ]; then
+				options="${options:+$options,}subvol=@var"
+			fi
+			if [ $mountpoint = /var/log ]; then
+                                options="${options:+$options,}subvol=@varlog"
+                        fi
+			if [ $mountpoint = /tmp ]; then
+				options="${options:+$options,}subvol=@tmp"
+			fi
+			if [ $mountpoint = /srv ]; then
+                                options="${options:+$options,}subvol=@srv"
+                        fi
+			if [ $mountpoint = /opt ]; then
+                                options="${options:+$options,}subvol=@opt"
+                        fi
+			if [ $mountpoint = /usr ]; then
+                                options="${options:+$options,}subvol=@usr"
+                        fi
+			if [ $mountpoint = /usr/share ]; then
+                                options="${options:+$options,}subvol=@usrshare"
+                        fi
+			if [ $mountpoint = /usr/local ]; then
+                                options="${options:+$options,}subvol=@usrlocal"
+                        fi
+			echo $path $mountpoint $filesystem $options 0 0
 			;;
 		esac
 	done
--- partman-btrfs-54.orig/mount.d/btrfs
+++ partman-btrfs-54/mount.d/btrfs
@@ -11,19 +11,90 @@ pass=$6
 
 case $type in
     btrfs)
-	# import workaround from Kali's partman-btrfs commit:7f43d2c
 	options="${options%,subvol=*}"
 	#for removing the option subvol,when thats the only option
 	#eg: options=="subvol=@", no comma present
 	options="${options%subvol=*}"
-	mount -t btrfs ${options:+-o "$options"} $fs /target"$mp" || exit 1
-	if [ $mp = / ]; then
-	    btrfs subvolume create /target$mp/@rootfs
-	    chmod 755 /target$mp/@rootfs
-	    umount /target$mp
-	    options="${options:+$options,}subvol=@rootfs"
-	    mount -t btrfs -o $options $fs /target$mp
-	fi
+	mount -t btrfs ${options:+-o "$options"} $fs /target$mp || exit 1
+	case $mp in
+	    /)
+		btrfs subvolume create /target$mp/@
+		chmod 755 /target$mp/@
+		umount /target$mp
+		options="${options:+$options,}subvol=@"
+		mount -t btrfs -o $options $fs /target$mp
+		;;
+	    /boot)
+                btrfs subvolume create /target$mp/@boot
+                chmod 755 /target$mp/@boot
+                umount /target$mp
+                options="${options:+$options,}subvol=@boot"
+                mount -t btrfs -o $options $fs /target$mp
+                ;;
+	    /home)
+		btrfs subvolume create /target$mp/@home
+		chmod 755 /target$mp/@home
+		umount /target$mp
+		options="${options:+$options,}subvol=@home"
+		mount -t btrfs -o $options $fs /target$mp
+		;;
+	    /var)
+		btrfs subvolume create /target$mp/@var
+		chmod 755 /target$mp/@var
+		umount /target$mp
+		options="${options:+$options,}subvol=@var"
+		mount -t btrfs -o $options $fs /target$mp
+		;;
+	    /var/log)
+                btrfs subvolume create /target$mp/@varlog
+                chmod 755 /target$mp/@varlog
+                umount /target$mp
+                options="${options:+$options,}subvol=@varlog"
+                mount -t btrfs -o $options $fs /target$mp
+		;;
+	    /tmp)
+		btrfs subvolume create /target$mp/@tmp
+		chmod 755 /target$mp/@tmp
+		umount /target$mp
+		options="${options:+$options,}subvol=@tmp"
+		mount -t btrfs -o $options $fs /target$mp
+		;;
+	    /srv)
+                btrfs subvolume create /target$mp/@srv
+                chmod 755 /target$mp/@srv
+                umount /target$mp
+                options="${options:+$options,}subvol=@srv"
+                mount -t btrfs -o $options $fs /target$mp
+		;;
+	    /opt)
+                btrfs subvolume create /target$mp/@opt
+                chmod 755 /target$mp/@opt
+                umount /target$mp
+                options="${options:+$options,}subvol=@opt"
+                mount -t btrfs -o $options $fs /target$mp
+		;;
+	    /usr)
+                btrfs subvolume create /target$mp/@usr
+                chmod 755 /target$mp/@usr
+                umount /target$mp
+                options="${options:+$options,}subvol=@usr"
+                mount -t btrfs -o $options $fs /target$mp
+                ;;
+	    /usr/share)
+                btrfs subvolume create /target$mp/@usrshare
+                chmod 755 /target$mp/@usrshare
+                umount /target$mp
+                options="${options:+$options,}subvol=@usrshare"
+                mount -t btrfs -o $options $fs /target$mp
+                ;;
+            /usr/local)
+                btrfs subvolume create /target$mp/@usrlocal
+                chmod 755 /target$mp/@usrlocal
+                umount /target$mp
+                options="${options:+$options,}subvol=@usrlocal"
+                mount -t btrfs -o $options $fs /target$mp
+                ;;
+	esac
 	echo "umount /target$mp"
 	exit 0
 	;;
